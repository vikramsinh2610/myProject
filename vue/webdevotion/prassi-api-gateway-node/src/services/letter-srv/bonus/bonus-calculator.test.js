const BonusLetter = require('./bonus-letter');
const { calculateBonuses } = require('./bonus-calculator');

describe('Bonus Calculator', () => {
  test('should calculate BonusLetter with Montly Prepaid Conditioned and Monthly Guaranteed', () => {
    const letter = new BonusLetter({
      _id: '01-2020',
      status: 'wip',
      promoterId: 'f34g3df-f3c3x2ed-d232r4g5h5-d21s',
      promoterSerialNumber: 'AZ123456',
      promoterDisplayName: 'Pippo De Peppis',
      type: 'rappel',
      description: 'Bla bla bla',
      signatureDate: new Date(),
      attachmentIds: ['a', 'b'],
      paymentFrequency: 'one-time',
      conditionedBonuses: [
        {
          amount: 1000,
          paymentTime: 'prepayment',
          productivePeriodPaymentDelay: 0,
          maxRecoveryPercentage: 10000,
          conditions: [
            {
              targets: [
                { kpi: { _id: 'KPI1', value: 500 }, targetValue: 500, weightPercentage: 5000 },
                { kpi: { _id: 'KPI2', value: 800 }, targetValue: 800, weightPercentage: 5000 },
              ],
            },
          ],
        },
      ],
      guaranteedBonuses: [
        {
          amount: 2000,
          paymentTime: 'payment-delayed',
          absorbability: {
            directProductionPercentage: 0,
            indirectProductionPercentage: 100,
          },
        },
      ],
      fromProductivePeriodYear: 2018,
      fromProductivePeriodMonth: 9,
      toProductivePeriodYear: 2019,
      toProductivePeriodMonth: 1,
      willExpireDate: new Date('2019-01-31T00:00:00.000Z'),
    });

    expect(calculateBonuses(letter, 2018, 9)).toMatchSnapshot('A');
    expect(calculateBonuses(letter, 2018, 10)).toMatchSnapshot('B');
    expect(calculateBonuses(letter, 2019, 1)).toMatchSnapshot('C');
  });

  test('should calculate BonusLetter with Payment Delayed Conditioned with delay', () => {
    const letter = new BonusLetter({
      _id: '01-2020',
      status: 'wip',
      promoterId: 'f34g3df-f3c3x2ed-d232r4g5h5-d21s',
      promoterSerialNumber: 'AZ123456',
      promoterDisplayName: 'Pippo De Peppis',
      type: 'rappel',
      description: 'Bla bla bla',
      signatureDate: new Date(),
      attachmentIds: ['a', 'b'],
      paymentFrequency: 'one-time',
      conditionedBonuses: [
        {
          amount: 1000,
          paymentTime: 'payment-delayed',
          productivePeriodPaymentDelay: 1,
          maxRecoveryPercentage: 10000,
          conditions: [
            {
              targets: [
                { kpi: { _id: 'KPI1', value: 500 }, targetValue: 500, weightPercentage: 5000 },
                { kpi: { _id: 'KPI2', value: 800 }, targetValue: 800, weightPercentage: 5000 },
              ],
            },
          ],
        },
      ],
      guaranteedBonuses: [],
      fromProductivePeriodYear: 2018,
      fromProductivePeriodMonth: 9,
      toProductivePeriodYear: 2019,
      toProductivePeriodMonth: 1,
      willExpireDate: new Date('2019-01-31T00:00:00.000Z'),
    });

    expect(calculateBonuses(letter, 2019, 1)).toMatchSnapshot('A');
    expect(calculateBonuses(letter, 2019, 2)).toMatchSnapshot('B');
  });

  test('should calculate BonusLetter with One Time Delayed Prepaid (failing) and One Time Guaranteed', () => {
    const letter = new BonusLetter({
      _id: '01-2020',
      status: 'wip',
      promoterId: 'f34g3df-f3c3x2ed-d232r4g5h5-d21s',
      promoterSerialNumber: 'AZ123456',
      promoterDisplayName: 'Pippo De Peppis',
      type: 'rappel',
      description: 'Bla bla bla',
      signatureDate: new Date(),
      attachmentIds: ['a', 'b'],
      paymentFrequency: 'one-time',
      conditionedBonuses: [
        {
          amount: 1000,
          paymentTime: 'prepayment',
          productivePeriodPaymentDelay: 0,
          maxRecoveryPercentage: 10000,
          conditions: [
            {
              targets: [
                { kpi: { _id: 'KPI1', value: 500 }, targetValue: 500, weightPercentage: 0 },
                { kpi: { _id: 'KPI2', value: 800 }, targetValue: 800, weightPercentage: 0 },
              ],
            },
          ],
        },
      ],
      guaranteedBonuses: [],
      fromProductivePeriodYear: 2018,
      fromProductivePeriodMonth: 9,
      toProductivePeriodYear: 2019,
      toProductivePeriodMonth: 1,
      willExpireDate: new Date('2019-01-31T00:00:00.000Z'),
    });

    expect(calculateBonuses(letter, 2019, 1)).toMatchSnapshot('A');
  });

  test('should calculate BonusLetter with One Time Delayed Prepaid (failing) and progressive recovery', () => {
    const letter = new BonusLetter({
      _id: '01-2020',
      status: 'wip',
      promoterId: 'f34g3df-f3c3x2ed-d232r4g5h5-d21s',
      promoterSerialNumber: 'AZ123456',
      promoterDisplayName: 'Pippo De Peppis',
      type: 'rappel',
      description: 'Bla bla bla',
      signatureDate: new Date(),
      attachmentIds: ['a', 'b'],
      paymentFrequency: 'monthly',
      conditionedBonuses: [
        {
          amount: 1000,
          paymentTime: 'prepayment',
          productivePeriodPaymentDelay: 0,
          maxRecoveryPercentage: 3000,
          conditions: [
            {
              targets: [
                { kpi: { _id: 'KPI1', value: 0 }, targetValue: 500, weightPercentage: 0 },
                { kpi: { _id: 'KPI2', value: 0 }, targetValue: 800, weightPercentage: 0 },
              ],
            },
          ],
        },
      ],
      guaranteedBonuses: [],
      fromProductivePeriodYear: 2018,
      fromProductivePeriodMonth: 9,
      toProductivePeriodYear: 2019,
      toProductivePeriodMonth: 1,
      willExpireDate: new Date('2019-01-31T00:00:00.000Z'),
    });

    expect(calculateBonuses(letter, 2019, 1)).toMatchSnapshot('A');
    expect(calculateBonuses(letter, 2019, 2)).toMatchSnapshot('B');
    expect(calculateBonuses(letter, 2019, 3)).toMatchSnapshot('C');
    expect(calculateBonuses(letter, 2019, 4)).toMatchSnapshot('D');
  });
});
