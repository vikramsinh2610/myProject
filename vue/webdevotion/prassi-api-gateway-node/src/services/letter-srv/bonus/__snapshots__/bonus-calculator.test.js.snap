// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bonus Calculator should calculate BonusLetter with Montly Prepaid Conditioned and Monthly Guaranteed: A 1`] = `
Array [
  ComputedBonus {
    "absorbability": Absorbability {
      "directProductionPercentage": 0,
      "indirectProductionPercentage": 0,
    },
    "accruedAmount": 1000,
    "expireDate": 2019-01-31T00:00:00.000Z,
    "invoiceDescription": null,
    "kpi": Array [
      Object {
        "targets": Array [
          Object {
            "kpi": Object {
              "_id": "KPI1",
              "value": 500,
            },
            "targetValue": 500,
            "weightPercentage": 5000,
          },
          Object {
            "kpi": Object {
              "_id": "KPI2",
              "value": 800,
            },
            "targetValue": 800,
            "weightPercentage": 5000,
          },
        ],
      },
    ],
    "targetAmount": 1000,
    "targetComplete": true,
    "targetPercentage": 10000,
    "type": "conditioned+rappel+prepayment",
  },
]
`;

exports[`Bonus Calculator should calculate BonusLetter with Montly Prepaid Conditioned and Monthly Guaranteed: B 1`] = `Array []`;

exports[`Bonus Calculator should calculate BonusLetter with Montly Prepaid Conditioned and Monthly Guaranteed: C 1`] = `
Array [
  ComputedBonus {
    "absorbability": Object {
      "directProductionPercentage": 0,
      "indirectProductionPercentage": 100,
    },
    "accruedAmount": 2000,
    "expireDate": 2019-01-31T00:00:00.000Z,
    "invoiceDescription": null,
    "kpi": Array [],
    "targetAmount": 2000,
    "targetComplete": true,
    "targetPercentage": 10000,
    "type": "guaranteed+one-time",
  },
]
`;

exports[`Bonus Calculator should calculate BonusLetter with One Time Delayed Prepaid (failing) and One Time Guaranteed: A 1`] = `
Array [
  ComputedBonus {
    "absorbability": Absorbability {
      "directProductionPercentage": 0,
      "indirectProductionPercentage": 0,
    },
    "accruedAmount": -1000,
    "expireDate": 2019-01-31T00:00:00.000Z,
    "invoiceDescription": null,
    "kpi": Array [],
    "targetAmount": 1000,
    "targetComplete": false,
    "targetPercentage": 10000,
    "type": "conditioned+rappel+recovery",
  },
]
`;

exports[`Bonus Calculator should calculate BonusLetter with One Time Delayed Prepaid (failing) and progressive recovery: A 1`] = `
Array [
  ComputedBonus {
    "absorbability": Absorbability {
      "directProductionPercentage": 0,
      "indirectProductionPercentage": 0,
    },
    "accruedAmount": -4000,
    "expireDate": 2019-01-31T00:00:00.000Z,
    "invoiceDescription": null,
    "kpi": Array [],
    "targetAmount": 5000,
    "targetComplete": false,
    "targetPercentage": 3000,
    "type": "conditioned+rappel+recovery",
  },
]
`;

exports[`Bonus Calculator should calculate BonusLetter with One Time Delayed Prepaid (failing) and progressive recovery: B 1`] = `Array []`;

exports[`Bonus Calculator should calculate BonusLetter with One Time Delayed Prepaid (failing) and progressive recovery: C 1`] = `Array []`;

exports[`Bonus Calculator should calculate BonusLetter with One Time Delayed Prepaid (failing) and progressive recovery: D 1`] = `Array []`;

exports[`Bonus Calculator should calculate BonusLetter with Payment Delayed Conditioned with delay: A 1`] = `Array []`;

exports[`Bonus Calculator should calculate BonusLetter with Payment Delayed Conditioned with delay: B 1`] = `
Array [
  ComputedBonus {
    "absorbability": Absorbability {
      "directProductionPercentage": 0,
      "indirectProductionPercentage": 0,
    },
    "accruedAmount": 1000,
    "expireDate": 2019-01-31T00:00:00.000Z,
    "invoiceDescription": null,
    "kpi": Array [
      Object {
        "targets": Array [
          Object {
            "kpi": Object {
              "_id": "KPI1",
              "value": 500,
            },
            "targetValue": 500,
            "weightPercentage": 5000,
          },
          Object {
            "kpi": Object {
              "_id": "KPI2",
              "value": 800,
            },
            "targetValue": 800,
            "weightPercentage": 5000,
          },
        ],
      },
    ],
    "targetAmount": 1000,
    "targetComplete": true,
    "targetPercentage": 10000,
    "type": "conditioned+rappel+payment-delayed",
  },
]
`;
